### PR TITLE
Rework the "picker" results to correctly manage files

### LINKED ISSUE #1477: [Enhancement] FilePicker provide file path readable by System.IO.File
## Summary
Actually in preview version, when I pick file from storage, picker return file path which is not compatible with file system format.

## Intended Use Case
The case is, I pick file, store path, and next time, I get file path and read stream from selected path. Now I must convert path to file system format.

### LINKED ISSUE #1481: [Bug] Using MediaPicker.CapturePhotoAsync(); Gets Error
### Description
when using **MediaPicker.CapturePhotoAsync();** get error "Object reference not set to an instance of an object"

I got this error when testing on **Android**.  

### Steps to Reproduce
            var result = await MediaPicker.CapturePhotoAsync();
            var stream = await result.OpenReadAsync();
            imgDocImage.Source = ImageSource.FromStream(() => stream);
just when running first line after shot the photo on phone gets the error :  "Object reference not set to an instance of an object"
on the first line, I mean before get a value in "result".

Test on Xiaomi note 9 (Android 10)

1. 
2. 
3. 

### Basic Information

- Version with issue: 1.6.0 pre2
- Last known good version:
- IDE: Visual Studio 2019
- Platform Target Frameworks: 
  - iOS:  
  - Android:  

- Android Support Library Version: 
- Nuget Packages: Xamarin.Essentials  1.6.0-pre2
- Affected Devices:   Xiaomi Redmi note 9   (Android 10)

### LINKED ISSUE #1483: [Bug] [FilePicker] [Android] Exception is thrown when picking some files
### Description
`Java.IO.FileNotFoundException: File is virtual` is thrown when pciking some files from Google drive. Not sure what are those virtual files exactly, but probably some file entities on Google Drive which have some dynamic content (just a pure guess).
### Steps to Reproduce

1. Have such "virtual file" on your google drive
2. Try to pick it
3. Call OpenReadAsync on FileResult

### Expected Behavior
Either throw Net exception (and not a Java one) or pick the file (if it's possible to pick those virtual files at all).
### Actual Behavior
Throws Java exception
### Basic Information

- Version with issue: 1.6.0-pre2
- Last known good version:
- IDE: VS 2019
- Platform Target Frameworks: 
   - Android:  11.0
- Affected Devices: Tested on Google Pixel 3a

### LINKED ISSUE #1537: [Bug] FilePicker permission for storage read doesn't work as intended on Android
### Description
Recently I started to use Xamarin.Essentials FilePicker for picking files from the file system. When using the FilePicker on Android (Android 10) I noticed that the permission prompt for the storage read permission didn't work properly: When choosing "Deny" from the prompt the file system is still opened, even though it shouldn't. From reading the source code (https://github.com/xamarin/Essentials/blob/main/Xamarin.Essentials/FilePicker/FilePicker.android.cs) it seems that the permission is only asked, but nothing is done with the result (line 19): await Permissions.RequestAsync<Permissions.StorageRead>();. 

### Steps to Reproduce

1. Bind the FilePicker PlatformPickAsync method to a button.
2. Run the application on an Android device (I did this on Android 10).
3. Press the button that has PlatformPickAsync  bound to
4. From the permission prompt select "Deny"

### Expected Behavior
File system doesn't open.

### Actual Behavior
File system opens, even though it shouldn't have the right permission. User is still able to navigate the file system and pick files from it.

### Basic Information

- Version with issue: Xamarin.Essentials.1.6.0-pre3
- Last known good version:
- IDE: Visual Studio 2019 16.8.1
- Platform Target Frameworks: 
  - Android: 10
- Affected Devices:
  - Samsung SM-G973F running Android 10 / One UI version 2.5
- Nuget Packages: Xamarin.Essentials.1.6.0-pre3

### Workaround

I was able to create a workaround by prompting the permission alert myself before using the FilePicker class' PlatformPickAsync method by using the Xamarin.Essentials Permission class. This didn't prompt a second permission alert from the FilePicker class on Android 10/Samsung SM-G973F, so it fit for my need at the moment.

### LINKED ISSUE #1538: [Bug] FilePicker stopped working on Android
### Description
Selecting any file on Android does not work.
### Steps to Reproduce

1. Start selecting the file. File picker will open
2. Pick the file which resides on the phone (to be sure that its binary payload is indeed available for the app). `await fileData.OpenReadAsync()`
3. Notice the exception

```
Java.Lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaDocumentsProvider uri content://com.android.providers.media.documents/document/video%3A28 from pid=10506, uid=10265 requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
  at Java.Interop.JniEnvironment+InstanceMethods.CallNonvirtualObjectMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x0008e] in <89755ea61d9c4ae0a40ce90b872c9e2d>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeNonvirtualObjectMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x0001f] in <89755ea61d9c4ae0a40ce90b872c9e2d>:0 
  at Android.Content.ContentResolver.OpenInputStream (Android.Net.Uri uri) [0x00031] in <4658d344ef4b4a429f9058200bb1b31c>:0 
  at Xamarin.Essentials.FileBase.PlatformOpenReadAsync () [0x0001e] in D:\a\1\s\Xamarin.Essentials\FileSystem\FileSystem.android.cs:119 
  at Xamarin.Essentials.FileBase.OpenReadAsync () [0x00000] in D:\a\1\s\Xamarin.Essentials\FileSystem\FileSystem.shared.cs:105 
  at MyClass.uploadFile (Xamarin.Essentials.FileResult fileData) [0x001ee] in D:\MyClass.cs:552 
  --- End of managed Java.Lang.SecurityException stack trace ---
java.lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaDocumentsProvider uri content://com.android.providers.media.documents/document/video%3A28 from pid=10506, uid=10265 requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
	at android.os.Parcel.createExceptionOrNull(Parcel.java:2373)
	at android.os.Parcel.createException(Parcel.java:2357)
	at android.os.Parcel.readException(Parcel.java:2340)
	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:190)
	at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel(DatabaseUtils.java:153)
	at android.content.ContentProviderProxy.openTypedAssetFile(ContentProviderNative.java:781)
	at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1983)
	at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:1798)
	at android.content.ContentResolver.openInputStream(ContentResolver.java:1475)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:30)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7656)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

### Expected Behavior
Should be able to pick the file

### Actual Behavior
Cannot pick the file

### Basic Information

- Version with issue: 1.6.0-pre4
- Last known good version: 1.6.0-pre3
- IDE: VS 2019 16.8.0
- Platform Target Frameworks: 
  
  - Android:  11
 
- Affected Devices: Pixel3a, Android 11

### Workaround
Downgrade to 1.6.0-pre3

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change ###

This PR solves a few issues.

* The URI to the picked file is **only valid as long as the calling Activity is alive**.  
   Because we use an intermediate activity, this means that the URI almost immediately becomes invalid.  
   To solve for this, we have to resolve the physical path of the picked file. Many providers have a way to retrieve the physical path, or the path is known based on the type of content URI.
* Some picked files **cannot be resolve to a physical path**.  
   This might be because there are no permissions, or the file only exists as a stream from a file provider.  
   In this case, we need to copy the file to a local location.
* Some picked **files are virtual*.   
   This might be a Google docs spreadsheet which is not a real file format.
   To solve this, we detect the virtual and then ask the provider what formats are supported. We then use that format and open the stream.

This PR also changes a few things in the code that overlap with the fixes.

* Removed the usage of magic strings to a static list of mime types and extensions.
* Made use of the `IsIntentSupported` method on Android.
* Switched to `List<T>` for `ShareMultipleFilesRequest.Files` for consistency.

### Bugs Fixed ###

- Fixes #1538
- Fixes #1477
- Fixes #1483
- Fixes #1537
- Fixes #1481

### API Changes ###

None visible.

### Behavioral Changes ###

This PR will make sure all the `FileResult` instances point to a physical file location after a picker operation.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of `main` at time of PR
- [x] Changes adhere to coding standard
- [x] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

**Review by pictos — APPROVED** [review]
That looks good, I have just one comment, but is more a suggestion than a request.

File: Xamarin.Essentials/FileSystem/FileSystem.android.cs [hunk 1]
```diff
@@ -31,67 +31,299 @@ static Task<Stream> PlatformOpenAppPackageFileAsync(string filename)
...
+            }
+
+            if (uri.Scheme.Equals("content", StringComparison.OrdinalIgnoreCase))
+            {
+                var resolved = ResolveContentPath(uri);
+                if (File.Exists(resolved))
+                    return resolved;
+            }
+            else if (uri.Scheme.Equals("file", StringComparison.OrdinalIgnoreCase))
+            {
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply Xamarin.Essentials/FileSystem/FileSystem.android.cs:89]: Yup. You got me. I am always pestering people abut these magic strings. I will hang my head in shame and walk to the nearest corner.

[the remark this answers is not in the document]
Comment (pictos) [reply Xamarin.Essentials/FileSystem/FileSystem.android.cs:89]: It's ok that you like to type the same string over and over again 🤣

File: Samples/Samples.Android/Properties/AndroidManifest.xml [hunk 2]
```diff
@@ -13,26 +13,35 @@
...
 		<intent>
 			<action android:name="android.intent.action.VIEW" />
 			<data android:scheme="smsto" />
 		</intent>
+		<!-- PhoneDialer -->
 		<intent>
 			<action android:name="android.intent.action.DIAL" />
 			<data android:scheme="tel" />
 		</intent>
+		<!-- MediaPicker -->
```
Comment (jamesmontemagno) [inline Samples/Samples.Android/Properties/AndroidManifest.xml:41]: This needs to be added to documentation :)

**Pull request merged**